File: obops/obops-facade/src/main/java/com/oceanbase/ocp/obops/tenant/model/Tenant.java
/*
 * Copyright (c) 2023 OceanBase
 * OCP Express is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

package com.oceanbase.ocp.obops.tenant.model;

import java.time.OffsetDateTime;

import com.oceanbase.ocp.core.ob.tenant.TenantMode;
import com.oceanbase.ocp.core.ob.tenant.TenantStatus;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Tenant {

    @Deprecated
    private Long id;

    private String name;

    private Long obTenantId;

    private TenantMode mode;

    private String clusterName;

    private Long obClusterId;

    private String obVersion;

    private OffsetDateTime createTime;

    private String primaryZone;

    private String zoneList;

    private String locality;

    private TenantStatus status;

    private Boolean locked;

    private Boolean readonly;

    private String description;
}

File: frontend/src/util/tenant.ts
/*
 * Copyright (c) 2023 OceanBase
 * OCP Express is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

import { formatMessage } from '@/util/intl';
import { min } from 'lodash';

// 获取集群各个 Zone 中 OBServer 数量的最小值，用于限制 OB 4.0 版本中可设置 Unit 的最大数量
export function getMinServerCount(zones: API.Zone[]) {
  if (zones?.length > 0) {
    return min(zones.map(zone => zone?.servers?.length || 0));
  }
  return 0;
}

// 根据  unitSpecLimit 说明不推荐使用的原因
export function getUnitSpecLimitText(unitSpecLimit: API.UnitSpecUnitSpecLimit) {
  const { memoryLowerLimit, cpuLowerLimit } = unitSpecLimit;
  if (memoryLowerLimit) {
    return formatMessage(
      {
        id: 'ocp-express.src.util.tenant.OnlyUnitSpecificationsWithAMemoryOfNo',
        defaultMessage: '仅可选择内存不小于 {memoryLowerLimit}G 的 Unit 规格',
      },
      { memoryLowerLimit: memoryLowerLimit }
    );
  }
  if (cpuLowerLimit) {
    return formatMessage(
      {
        id: 'ocp-express.src.util.tenant.OnlyUnitSpecificationsWithCpuNoLessThan',
        defaultMessage: '仅可选择 CPU 不小于 {cpuLowerLimit}C 的 Unit 规格',
      },
      { cpuLowerLimit: cpuLowerLimit }
    );
  }
}
